File: src/routeros/history.js
'use strict';
/**
 * Undo and redo for resource writes — the part that can be reasoned about
 * without a socket.
 *
 * Every write the engine performs is recorded as a pair of operations: the one
 * that was done, and the one that reverses it. Undo runs the reverse, redo runs
 * the forward again. Nothing here talks to a router; index.js applies these.
 *
 * ── Positions are ANCHORS, never indexes ──────────────────────────────────
 *
 * For an ordered resource — the firewall — restoring a row means putting it
 * back where it was, and "where it was" is recorded as *the id of the row it
 * sat immediately before*, or null for the end of the table. An ordinal would
 * be wrong the moment anything else in the table moved, and the whole point of
 * undo is that time has passed since the action.
 *
 * That anchor can itself be deleted, and then the entry is unusable. It is
 * refused rather than approximated: putting a firewall rule back in roughly the
 * right place is worse than saying it cannot be done.
 *
 * ── What is deliberately not recorded ─────────────────────────────────────
 *
 * A secret. `rowValues()` never returns one, so a `before` cannot contain a
 * pre-shared key, and undoing an edit that changed one leaves the current key
 * alone rather than restoring a value this process never had.
 */

/** The verbs an entry can ask index.js to perform. */
const OPS = Object.freeze(['add', 'remove', 'set', 'move', 'enable', 'disable']);

const SEP = '\u0001';

/**
 * A short sentence for the button's tooltip: "undo delete of 192.0.2.0/24".
 *
 * The identity is a composite for some resources, so its separator is swapped
 * for something readable rather than shown raw.
 */
function label(resource, what, identity) {
  const name = String(identity || '').split(SEP).filter(Boolean).join(' ');
  const verb = { create: 'add', update: 'edit', delete: 'delete',
                 move: 'move', enable: 'enable', disable: 'disable' }[what] || what;
  return verb + ' of ' + (name || String(resource.label || '').toLowerCase());
}

/**
 * Record a completed write.
 *
 * `before` and `after` are resource-named values (what rowValues() produces),
 * not RouterOS rows. `anchorBefore` / `anchorAfter` are the id the row sat
 * before, and are only meaningful for an ordered resource.
 */
function buildEntry({ resource, what, id, identity, before, after, anchorBefore, anchorAfter }) {
  let forward, reverse;

  switch (what) {
    case 'create':
      forward = { op: 'add', values: after, anchor: anchorAfter };
      reverse = { op: 'remove', id };
      break;

    case 'delete':
      forward = { op: 'remove', id };
      // Re-adding gives the row a NEW id, so the anchor is how it finds its
      // place again. index.js writes the new id back into the other half.
      reverse = { op: 'add', values: before, anchor: anchorBefore };
      break;

    case 'update':
      forward = { op: 'set', id, values: after };
      reverse = { op: 'set', id, values: before };
      break;

    case 'move':
      forward = { op: 'move', id, anchor: anchorAfter };
      reverse = { op: 'move', id, anchor: anchorBefore };
      break;

    case 'enable':
      forward = { op: 'enable', id };
      reverse = { op: 'disable', id };
      break;

    case 'disable':
      forward = { op: 'disable', id };
      reverse = { op: 'enable', id };
      break;

    default:
      return null;                    // an action with no inverse is not recorded
  }

  return { resource: resource.key, what, identity: identity || '', forward, reverse,
           label: label(resource, what, identity) };
}

/**
 * Keep both halves pointing at the row that now exists.
 *
 * An `add` has no id until it runs, so applying one has to write the resulting
 * id back into the entry — otherwise the matching `remove` would have nothing
 * to address.
 */
function rebind(entry, id) {
  if (!entry) return;
  for (const op of [entry.forward, entry.reverse]) if (op && op.op !== 'add') op.id = id;
}

module.exports = { OPS, SEP, buildEntry, label, rebind };

File: internal/history/history.go
// Package history is undo and redo for resource writes — the part that can be
// reasoned about without a router.
//
// A port of src/routeros/history.js. Every write the server performs is recorded
// as a PAIR of operations: the one that was done, and the one that reverses it.
// Undo runs the reverse, redo runs the forward again. Nothing here talks to a
// router; internal/server applies these.
//
// Pure, and deliberately so — the same argument the guard modules make. Values
// in, entry out, no I/O, so every rule is testable without hardware.
//
// ── What is deliberately not recorded ────────────────────────────────────────
//
// A secret. RowValues never returns one, so a `before` cannot contain a
// pre-shared key, and undoing an edit that changed one leaves the current key
// alone rather than restoring a value this process never had.
//
// ── What this port does NOT carry yet, and why ───────────────────────────────
//
// The live module also handles `move`, `enable` and `disable`, and records a
// POSITION for an ordered resource as an ANCHOR — the id of the row it sat
// immediately before, never an ordinal, because an ordinal is wrong the moment
// anything else in the table moves and undo exists precisely because time has
// passed. An anchor that has itself been deleted makes the entry unusable, and
// the live app refuses it rather than approximating: putting a firewall rule
// back in roughly the right place is worse than saying it cannot be done.
//
// None of that is here, because nothing that reaches it is ported: no resource
// declares `Ordered`, and neither `res:action` nor a move path exists on this
// side. Writing it now would be untestable code guarding a case that cannot
// occur. It lands with the firewall, the first ordered resource in the queue —
// and `TestPortedFieldsMatchTheirLiveDeclarations` now compares `Ordered`
// against the live declaration, so that port cannot quietly skip it.
package history

import "strings"

// Sep joins the parts of a composite identity. Some resources identify a row by
// more than one field; the separator is swapped for a space before a human sees
// it rather than shown raw.
const Sep = "\x01"

// Op is one recorded operation, in the vocabulary RouterOS has verbs for.
type Op struct {
	// Op is add, set or remove.
	Op string
	// ID addresses an existing row. An `add` has none until it runs.
	ID string
	// Values are resource-named, as Validate takes them — never RouterOS rows.
	Values map[string]string
}

// Entry is a completed write and its inverse.
type Entry struct {
	Resource string
	What     string
	Identity string
	Forward  Op
	Reverse  Op
	Label    string
}

var verbs = map[string]string{
	"create": "add", "update": "edit", "delete": "delete",
}

// Label is the sentence the button's tooltip shows: "undo delete of
// 192.0.2.0/24". Falls back to the resource's own label when the row has no
// identity worth naming.
func Label(resourceLabel, what, identity string) string {
	parts := make([]string, 0, 2)
	for _, p := range strings.Split(identity, Sep) {
		if p != "" {
			parts = append(parts, p)
		}
	}
	name := strings.Join(parts, " ")
	verb, ok := verbs[what]
	if !ok {
		verb = what
	}
	if name == "" {
		name = strings.ToLower(resourceLabel)
	}
	return verb + " of " + name
}

// Build records a completed write.
//
// `before` and `after` are resource-named values — what RowValues produces —
// not RouterOS rows. A `what` with no inverse returns nil and is not recorded,
// which is how an unrecognised action fails: absent from the stack rather than
// present and wrong.
func Build(resourceKey, resourceLabel, what, id, identity string,
	before, after map[string]string) *Entry {
	var forward, reverse Op
	switch what {
	case "create":
		forward = Op{Op: "add", Values: after}
		reverse = Op{Op: "remove", ID: id}
	case "delete":
		forward = Op{Op: "remove", ID: id}
		// Re-adding gives the row a NEW id, so the server writes it back into
		// the other half once the add has run — see Rebind.
		reverse = Op{Op: "add", Values: before}
	case "update":
		forward = Op{Op: "set", ID: id, Values: after}
		reverse = Op{Op: "set", ID: id, Values: before}
	default:
		return nil
	}
	return &Entry{
		Resource: resourceKey, What: what, Identity: identity,
		Forward: forward, Reverse: reverse,
		Label: Label(resourceLabel, what, identity),
	}
}

// Rebind keeps both halves pointing at the row that now exists.
//
// An `add` has no id until it runs, so applying one has to write the resulting
// id back into the entry — otherwise the matching `remove` would have nothing to
// address.
func Rebind(e *Entry, id string) {
	if e == nil {
		return
	}
	if e.Forward.Op != "add" {
		e.Forward.ID = id
	}
	if e.Reverse.Op != "add" {
		e.Reverse.ID = id
	}
}
